File: SWExpertAcademy/H2268_removingBlocks.cpp
#include <iostream>
#include <vector>
using namespace std;

vector<vector<int>> v;
int i, j, time, cnt, r, c, flag;

void init(int R, int C)
{
    r = R;
    c = C;
    for(i = 0; i < c; i++){
        vector<int> tmp;
        v.push_back(tmp);
    }
    time = 0;
}

int dropBlocks(int mTimestamp, int mCol, int mLen)
{
    cnt += mLen;

    vector<vector<int>> v2;
    for(i = 0; i < c; i++){
        vector<int> tmp;
        for(j = 0; j < v[i].size(); j++){
            v[i][j] += mTimestamp - time;
            if(v[i][j] >= r){
                cnt -= 1;
            }
            else{
                tmp.push_back(v[i][j]);
            }
        }
        v2.push_back(tmp);
    }
    time = mTimestamp;

    v = v2;

    for(i = 0; i < mLen; i++){
        v[mCol + i].push_back(0);
    }


    // cout << cnt << '\n';
    // for(i = 0; i < c; i++){
    //     int flag = 0;
    //     for(j = 0; j < v[i].size(); j++){
    //         cout << v[i][j] << " ";
    //         flag = 1;
    //     }
    //     if(flag)
    //         cout << "\n";
    // }
    // cout << "\n";
    return cnt;
}

int removeBlocks(int mTimestamp)
{
    vector<vector<int>> v2;
    for(i = 0; i < c; i++){
        vector<int> tmp;
        flag = 1;
        for(j = 0; j < v[i].size(); j++){
            v[i][j] += mTimestamp - time;
            if(v[i][j] >= r){
                cnt -= 1;
            }
            else{
                if(flag){
                    cnt -= 1;
                    flag = false;
                }
                else{
                    tmp.push_back(v[i][j]);
                }
            }

        }
        v2.push_back(tmp);
    }

    v = v2;
    time = mTimestamp;

    // cout << cnt << '\n';
    // for(i = 0; i < c; i++){
    //     for(j = 0; j < v[i].size(); j++){
    //         cout << v[i][j] << " ";
    //     }
    //     cout << "\n";
    // }
    // cout << "\n";
    return cnt;
}


#ifndef _CRT_SECURE_NO_WARNINGS
#define _CRT_SECURE_NO_WARNINGS
#endif

#include <stdio.h>

void init(int R, int C);
int dropBlocks(int mTimestamp, int mCol, int mLen);
int removeBlocks(int mTimestamp);

#define CMD_INIT 100
#define CMD_DROP 200
#define CMD_REMOVE 300

static bool run()
{
    int query_num;
    scanf("%d", &query_num);

    int ret, ans;
    bool ok = false;

    for (int q = 0; q < query_num; q++)
    {
        int query;
        scanf("%d", &query);

        if (query == CMD_INIT)
        {
            int R, C;
            scanf("%d %d", &R, &C);
            init(R, C);
            ok = true;
        }
        else if (query == CMD_DROP)
        {
            int mTimestamp, mCol, mLen;
            scanf("%d %d %d", &mTimestamp, &mCol, &mLen);
            ret = dropBlocks(mTimestamp, mCol, mLen);
            scanf("%d", &ans);
            if (ans != ret)
            {
                ok = false;
            }
        }
        else if (query == CMD_REMOVE)
        {
            int mTimestamp;
            scanf("%d", &mTimestamp);
            ret = removeBlocks(mTimestamp);
            scanf("%d", &ans);
            if (ans != ret)
            {
                ok = false;
            }
        }

    }
    return ok;
}

int main()
{
    setbuf(stdout, NULL);
    // freopen("sample_input.txt", "r", stdin);
    int T, MARK;
    scanf("%d %d", &T, &MARK);

    for (int tc = 1; tc <= T; tc++)
    {
        int score = run() ? MARK : 0;
        printf("#%d %d\n", tc, score);
    }
    return 0;
}

File: SWExpertAcademy/H2268_removingBlocks.py
global time, cnt, r, c, flag, v

d = dict()
v = []

def init(R: int, C: int) -> None:
    global time, cnt, r, c, flag, v
    r = R
    c = C
    for i in range(r):
        v.append([])
    time = 0
    cnt = 0

    return


def dropBlocks(mTimestamp: int, mCol: int, mLen: int) -> int:
    global time, cnt, r, c, flag, v
            
    for i in d:
        tmp = []
        for j in range(len(d[i])):
            if d[i][j] + mTimestamp - time >= r:
                cnt -= 1
                continue
            else:
                d[i][j] += mTimestamp - time
                tmp.append(d[i][j])
        d[i] = tmp

    for i in range(mLen):
        if mCol + i in d:
            d[mCol + i].append(0)
        else:
            d[mCol + i] = [0]

    cnt += mLen

    time = mTimestamp


    #######




    for i in d:
        print(i, d[i])





    # for i in range(c):
    #     tmp = []
    #     for j in range(len(v[i])):
    #         v[i][j] += mTimestamp - time
    #         if v[i][j] >= r:
    #             cnt -= 1
    #         else:
    #             tmp.append(v[i][j])
    #     v2.append(tmp)
    # time = mTimestamp
    # v = copy.deepcopy(v2)

    # for i in range(mLen):
    #     v[mCol + i].append(0)

    return cnt


def removeBlocks(mTimestamp: int) -> int:
    global time, cnt, r, c, flag, v
            
    for i in d:
        for j in range(len(d[i])):
            if d[i][j] + mTimestamp - time >= r:
                cnt -= 1
                continue
            else:
                d[i][j] += mTimestamp - time

    for i in d:
        cnt -= 1
        d[i] = d[i][1:]
    
    time = mTimestamp




    #######




    for i in d:
        print(i, d[i])



    
        
    # v2 = []
    # for i in range(c):
    #     tmp = []
    #     flag = 1
    #     for j in range(len(v[i])):
    #         v[i][j] += mTimestamp - time
    #         if v[i][j] >= r:
    #             cnt -= 1
    #         else:
    #             if flag:
    #                 cnt -= 1
    #                 flag = 0
                
    #             else:
    #                 tmp.append(v[i][j])
    #     v2.append(tmp)

    # v = copy.deepcopy(v2)
    # time = mTimestamp

    return cnt




# import copy
# global time, cnt, r, c, flag, v

# v = []

# def init(R: int, C: int) -> None:
#     global i, j, time, cnt, r, c, flag, v
#     r = R
#     c = C
#     for i in range(c):
#         v.append([])
#     time = 0

#     return


# def dropBlocks(mTimestamp: int, mCol: int, mLen: int) -> int:
#     global i, j, time, cnt, r, c, flag, v
    
#     cnt += mLen
#     v2 = []
#     for i in range(c):
#         tmp = []
#         for j in range(len(v[i])):
#             v[i][j] += mTimestamp - time
#             if v[i][j] >= r:
#                 cnt -= 1
#             else:
#                 tmp.append(v[i][j])
#         v2.append(tmp)
#     time = mTimestamp
#     v = copy.deepcopy(v2)

#     for i in range(mLen):
#         v[mCol + i].append(0)

#     return cnt


# def removeBlocks(mTimestamp: int) -> int:
#     global i, j, time, cnt, r, c, flag, v

#     v2 = []
#     for i in range(c):
#         tmp = []
#         flag = 1
#         for j in range(len(v[i])):
#             v[i][j] += mTimestamp - time
#             if v[i][j] >= r:
#                 cnt -= 1
#             else:
#                 if flag:
#                     cnt -= 1
#                     flag = 0
                
#                 else:
#                     tmp.append(v[i][j])
#         v2.append(tmp)

#     v = copy.deepcopy(v2)
#     time = mTimestamp

#     return cnt


import sys

CMD_INIT = 100
CMD_DROP = 200
CMD_REMOVE = 300


def run():
    query = int(input())
    ok = False
    for i in range(query):
        input_iter = iter(input().split())
        cmd = int(next(input_iter))
        if cmd == CMD_INIT:
            R = int(next(input_iter))
            C = int(next(input_iter))
            init(R, C)
            ok = True
        elif cmd == CMD_DROP:
            mTimestamp = int(next(input_iter))
            mCol = int(next(input_iter))
            mLen = int(next(input_iter))
            ret = dropBlocks(mTimestamp, mCol, mLen)
            ans = int(next(input_iter))
            if ans != ret:
                ok = False
        elif cmd == CMD_REMOVE:
            mTimestamp = int(next(input_iter))
            ret = removeBlocks(mTimestamp)
            ans = int(next(input_iter))
            if ans != ret:
                ok = False
    return ok


if __name__ == '__main__':
    # sys.stdin = open('sample_input.txt', 'r')
    inputarray = input().split()
    TC = int(inputarray[0])
    MARK = int(inputarray[1])

    for testcase in range(1, TC + 1):
        score = MARK if run() else 0
        print("#%d %d" % (testcase, score), flush=True)
